FIX base theme initialization and theme bootup

ndelangen opened the pull request:
Issue: #5824 #5780 

creating a theme that derived off of `light` or `dark` default themes was impractical, and wasn't working quite as users were expecting it to behave.


## What I did
ADD ability for users to specify baseTheme, and the theme createFn will merge in the appropriate themeVars.

ndelangen committed:
ADD ability for users to specify baseTheme, and the theme createFn will merge in the appropriate themeVars

ndelangen committed:
FIX circular dependencies

shilman commented on lib/theming/src/themes/light.ts:
```diff
@@ -1,36 +1,10 @@
 import { create } from '../create';
-import { color, typography, background } from '../base';
+
+import { themeVars } from './dark-vars';
```
@ndelangen should this be from `light-vars`?

ndelangen replied:
indeed

shilman reviewed (changes requested):
@ndelangen please test the hell out of this?

tmeasday commented on lib/theming/src/themes/dark.ts:
```diff
@@ -1,36 +1,9 @@
-  inputBorder: 'rgba(0,0,0,.3)',
-  inputTextColor: color.lightest,
-  inputBorderRadius: 4,
+  ...themeVars,
```
Isn't this superfluous?

tmeasday replied:
As in, wouldn't `create({ base: 'dark' })` have the exact same effect, by definition.

tmeasday commented on lib/theming/src/create.ts:
```diff
@@ -110,76 +115,85 @@ const darkSyntaxColors = {
+    base: vars.base,
+    color: createColors(inherit),
+    background: {
+      app: p('appBg'),
```
I don't really understand why this isn't just `inherit.appBg`?

ndelangen replied:
hmm, yeah

tmeasday commented on lib/theming/src/create.ts:
```diff
@@ -110,76 +115,85 @@ const darkSyntaxColors = {
+}
+
+export const create = (vars: ThemeVar, rest?: Rest): Theme => {
+  const inherit: ThemeVar = { ...vars, ...(base[vars.base] || base.light), ...base.light };
```
Should this just be `themeVars` or something

tmeasday replied:
I mean this is just a naming question, not a big deal ;)

tmeasday reviewed (changes requested):
I feel like this code would benefit from some tests but I appreciate this is a fairly late change

shilman commented:
When I add this line to `offcial-storybook/config.js` the app hangs:

```
    theme: create({ base: 'dark', colorPrimary: '#FF4785', colorSecondary: '#1EA7FD' }),
```

ndelangen committed:
Merge branch 'next' into tech/improve-theme-creating

ndelangen committed:
RENAME ThemeVar to ThemeVars && REMOVE the pick & p function

ndelangen commented:
I'm writing tests @shilman

ndelangen committed:
FIX issue where the `base` property wouldn't be set if non was given

ndelangen committed:
CLEANUP the creation of default themes

ndelangen committed:
MOVE ThemeVars to base && FIX custom properties overriding known properties && FIX merge order of vars & defaults

ndelangen committed:
ADD tests

shilman commented:
@ndelangen A couple comments, both outside the scope of this PR:
- After testing out a buggy version, I had to do `localStorage.clear()` to get the theme to load. Seems like a bug, but maybe it's a corner case
- Seems like the theme serialization should be profiled.. pretty slow to load on my laptop

shilman reviewed (changes requested):
@ndelangen Maybe a stupid question:

```
    theme: create({ base: 'dark', colorPrimary: 'red', colorSecondary: 'green' }),
```

I did this in `official-storybook`. I see green in the resulting UI, but I don't see any red. Doesn't seem right to me?

ndelangen committed:
FIX a bug with animations not being stringify-able

ndelangen commented:
@shilman 

I'd like to remove the need to do theme serialization asap.

theme.color.primary seems to be very rarely used, @domyen came up with the naming and usage of colors.

domyen commented:
@ndelangen yea `theme.color.primary` is not used at all, happy for you to axe it! (the var was mirrored from storybook/frontpage, but we don't need it here)

shilman commented:
@domyen @ndelangen seems like an unfortunate choice of names then... I thought the point of this minimal theme var abstraction was to help us preserve backwards compatibility?

shilman commented:
We should be making minimal changes at this point though, so I don't really want to change it. Though it makes me cringe a bit.

domyen commented:
I’m talking about what’s exposed in the create() function “colorPrimary”.

We still want to have a theme.color.primary as a part of the larger theme
colors (it’s set to storybook pink).
On Mon, Mar 4, 2019 at 12:00 PM Michael Shilman <notifications@github.com>
wrote:

> @domyen <https://github.com/domyen> @ndelangen
> <https://github.com/ndelangen> seems like an unfortunate choice of names
> then... I thought the point of this minimal theme var abstraction was to
> help us preserve backwards compatibility?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/storybooks/storybook/pull/5843#issuecomment-469330431>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQE2S4UQiw8NenGNAS1cndCZ8UwS1NDks5vTVE4gaJpZM4bcCBU>
> .
>

ndelangen committed:
REFACTOR change the theme that users import & set to be ThemeVars (was Theme)

This way the object easy much easier and more performant to serialise & less trickery is needed to make it work well

ndelangen committed:
OPTIMIZE so the theme doesn't update every story change

ndelangen commented:
@shilman @tmeasday 

I did a refactor to ensure the least amount of data needed is transferred over the channel.

I ended up not 'creating' the full theme and then trying to transmit, instead it's much more efficient to transfer the `ThemeVars` across and construct the actual `Theme` in the manager.

This means no more hack to make sure the functions make it across, since is a plain flat object.

I know it's not what you were hoping for: it's a bigger change I wish was necessary.

ndelangen commented:
I wrote in such a way, users won't need to change anything.
As long as they are using `create()`..

ndelangen commented:
I also noticed that setOptions is called a ton, and it's was re-creating the theme on every navigation.

It'd be nice to fix this a bit better, but for now I improved it by `deepEqual(theme, updatedTheme)`.

ndelangen committed:
Merge branch 'next' into tech/improve-theme-creating

ndelangen committed:
FIX tests

ndelangen committed:
FIX incorrect optimization

ndelangen committed:
ADD special case for `barSelectedColor`, since it should be using `colorSecondary` as fallback

Commit:
FIX the location of the default theme on state

Commit:
Ensure the theme is insta-loaded from local storage

Commit:
FIX Don't retain persisted UI state

For #5857

shilman committed:
Merge pull request #5856 from storybooks/tech/load-initial-theme

Ensure the theme is insta-loaded from local storage

shilman committed:
Merge pull request #5858 from storybooks/tech/workaround-persisted-options-with-defaults

Tech/workaround persisted options with defaults
